backend/src/main/java/com/rota/api/StaffService.java

```diff
@@ -21,6 +24,40 @@
   @Autowired
   StaffRepository staffRepository;
 
+  /**
+   * Checks if the engagement has any overlap with the requested time frame.
+   * 
+   * @param start start time
+   * @param end end time
+   * @return true if engagement times have any overlap with requested times
+   */
+  private Predicate<Engagement> isEngagementBetween(Instant start, Instant end) {
+    Instant startTime = start == null ? Instant.MIN : start;
+    Instant endTime = end == null ? Instant.MAX : end;
+
+    return (Engagement engagement) -> !engagement.getEnd().isBefore(startTime)
+        && !engagement.getStart().isAfter(endTime);
+  }
+
+  /**
+   * Truncates the engagement, e.g. if if starts before the start time the function will cut off the
+   * time difference and set the engagement start time to requested start time.
+   */
+  private Function<Engagement, Engagement> truncateEngagement(Instant start, Instant end) {
```

Comment:
I like how you've taken these out of the lambda expressions, I didn't know you could do this in Java.

Reply:
Everything in java is a class, `lambda`s are no exception.
Under the hood lambda expressions are classes from `java.util.function`.

Not sure if this is the most OO way of handling this, I may have been corrupted by functional programming.

Reply:
I did some reading on it and some places were suggesting that a lambda shouldn't be too long otherwise it gets too complicated, so taking this logic out of there makes sense. If we needed to reuse `isEngagementBetween` in a different scenario a boolean might have been better, but we don't so I think this way is good!